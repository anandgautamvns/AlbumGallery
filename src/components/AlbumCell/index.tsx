import {useNavigation} from '@react-navigation/native';
import React, {memo, useCallback} from 'react';
import {NavProp} from '../../navigation/AppNavigation';
import {Album} from './../../services/type';

import {ImageSourcePropType} from 'react-native';
import styled from 'styled-components/native';
import {useOrientation} from '../../hooks/useOrientation';

interface Props {
  album: Album;
}

const AlbumCell: React.FC<Props> = ({album}) => {
  const navigation = useNavigation<NavProp<'Details'>>();
  const isPortrait = useOrientation();

  const handlePress = useCallback(() => {
    navigation.navigate('Details', {album});
  }, [navigation, album]);

  const imageUri = isPortrait ? album.artworkUrl60 : album.artworkUrl100;
  const validUri =
    imageUri && imageUri.startsWith('http') ? imageUri : 'fallback_image_url';

  return (
    <Card onPress={handlePress} testID="album-card">
      <Artwork
        testID="album-artwork"
        key={isPortrait ? 'portrait' : 'landscape'}
        source={{uri: validUri} as ImageSourcePropType}
      />
      <Row>
        <Title numberOfLines={1}>
          {album.collectionName || album.trackName}
        </Title>
        <Price>{`$${album.trackPrice}`}</Price>
      </Row>
      {album.artistName && <Artist>{album.artistName}</Artist>}
    </Card>
  );
};

export default memo(AlbumCell);

// Styled-components
const Card = styled.TouchableOpacity`
  flex: 1;
  margin: 8px;
  background-color: #fff;
  border-radius: 12px;
  elevation: 3;
  shadow-color: #000;
  shadow-opacity: 0.1;
  shadow-radius: 6px;
  padding-bottom: 10px;
`;

const Artwork = styled.Image`
  width: 100%;
  aspect-ratio: 1;
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
  resize-mode: cover;
`;

const Row = styled.View`
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding-horizontal: 5px;
  width: 100%;
`;

const Title = styled.Text`
  flex: 1;
  margin-right: 8px;
  font-size: 10px;
  font-weight: 800;
  color: #000;
`;

const Price = styled.Text`
  font-size: 10px;
  font-weight: bold;
  color: #000;
  flex-shrink: 0;
`;

const Artist = styled.Text`
  font-size: 10px;
  color: #666;
  padding-horizontal: 5px;
  margin-top: 4px;
`;
