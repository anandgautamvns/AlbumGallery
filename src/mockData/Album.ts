import {Album, AlbumResponse} from '../services/type';

export const generateMockAlbums = (count: number): Album[] => {
  const mockAlbums: Album[] = [];
  for (let i = 1; i <= count; i++) {
    mockAlbums.push({
      wrapperType: 'track',
      kind: 'feature-movie',
      collectionId: 1000000000 + i,
      trackId: 2000000000 + i,
      artistName: `Artist ${i}`,
      collectionName: `Collection ${i}`,
      trackName: `Track ${i}`,
      collectionCensoredName: `Censored Collection ${i}`,
      trackCensoredName: `Censored Track ${i}`,
      collectionArtistId: 3000000000 + i,
      collectionArtistViewUrl: `https://example.com/artist/${i}`,
      collectionViewUrl: `https://example.com/collection/${i}`,
      trackViewUrl: `https://example.com/track/${i}`,
      previewUrl: `https://example.com/preview/${i}.mp4`,
      artworkUrl30: `https://example.com/artwork30/${i}.jpg`,
      artworkUrl60: `https://example.com/artwork60/${i}.jpg`,
      artworkUrl100: `https://example.com/artwork100/${i}.jpg`,
      collectionPrice: parseFloat((i * 20).toFixed(2)),
      trackPrice: parseFloat((i * 20).toFixed(2)),
      trackRentalPrice: parseFloat((i * 5).toFixed(2)),
      collectionHdPrice: parseFloat((i * 25).toFixed(2)),
      trackHdPrice: parseFloat((i * 25).toFixed(2)),
      trackHdRentalPrice: parseFloat((i * 5).toFixed(2)),
      releaseDate: new Date(
        Date.now() - Math.floor(i * 10000000000),
      ).toISOString(),
      collectionExplicitness: 'notExplicit',
      trackExplicitness: 'notExplicit',
      discCount: 1,
      discNumber: 1,
      trackCount: 10,
      trackNumber: i,
      trackTimeMillis: Math.floor(i * 10000000),
      country: 'USA',
      currency: 'USD',
      primaryGenreName: 'Action & Adventure',
      contentAdvisoryRating: 'PG-13',
      shortDescription: `Short description for track ${i}`,
      longDescription: `Long description for track ${i}. This is a detailed description of the track.`,
      hasITunesExtras: i > 0.5,
    });
  }
  return mockAlbums;
};

export const generateMockAlbumList = (count: number): AlbumResponse => {
  return {
    resultCount: count,
    results: generateMockAlbums(count),
  };
};

// Example usage:
export const mockAlbumsData = generateMockAlbums(5);
export const mockAlbumList = generateMockAlbumList(5);
