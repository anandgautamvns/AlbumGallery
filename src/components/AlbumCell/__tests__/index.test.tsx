import {useNavigation} from '@react-navigation/native';
import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import AlbumCell from '..';
import {mockAlbumsData} from '../../../mockData/Album';

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));

jest.mock('../../../hooks/useOrientation', () => ({
  useOrientation: jest.fn(() => true), // Mock portrait mode
}));

describe('AlbumCell Component', () => {
  const mockNavigation = {navigate: jest.fn()};
  const mockAlbum = mockAlbumsData[0];

  beforeEach(() => {
    (useNavigation as jest.Mock).mockReturnValue(mockNavigation);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const {toJSON} = render(<AlbumCell album={mockAlbum} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render album details correctly', () => {
    const {getByText, getByTestId} = render(<AlbumCell album={mockAlbum} />);

    expect(getByText('Collection 1')).toBeTruthy();
    expect(getByText('$20')).toBeTruthy();
    expect(getByText('Artist 1')).toBeTruthy();
    expect(getByTestId('album-artwork').props.source.uri).toBe(
      'https://example.com/artwork60/1.jpg',
    );
  });

  it('should navigate to the Details screen when pressed', () => {
    const {getByTestId} = render(<AlbumCell album={mockAlbum} />);

    fireEvent.press(getByTestId('album-card'));
    expect(mockNavigation.navigate).toHaveBeenCalledWith('Details', {
      album: mockAlbum,
    });
  });

  it('should use fallback image if artwork URL is invalid', () => {
    const invalidAlbum = {...mockAlbum, artworkUrl60: ''};
    const {getByTestId} = render(<AlbumCell album={invalidAlbum} />);

    expect(getByTestId('album-artwork').props.source.uri).toBe(
      'fallback_image_url',
    );
  });
});
