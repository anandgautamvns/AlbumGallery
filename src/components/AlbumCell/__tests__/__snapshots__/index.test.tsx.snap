// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumCell Component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#fff",
      "borderBottomLeftRadius": 12,
      "borderBottomRightRadius": 12,
      "borderTopLeftRadius": 12,
      "borderTopRightRadius": 12,
      "elevation": 3,
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
      "marginBottom": 8,
      "marginLeft": 8,
      "marginRight": 8,
      "marginTop": 8,
      "opacity": 1,
      "paddingBottom": 10,
      "shadowColor": "#000",
      "shadowOpacity": 0.1,
      "shadowRadius": 6,
    }
  }
  testID="album-card"
>
  <Image
    source={
      {
        "uri": "https://example.com/artwork60/1.jpg",
      }
    }
    style={
      {
        "aspectRatio": 1,
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "resizeMode": "cover",
        "width": "100%",
      }
    }
    testID="album-artwork"
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 5,
        "width": "100%",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#000",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontSize": 10,
          "fontWeight": "800",
          "marginRight": 8,
        }
      }
    >
      Collection 1
    </Text>
    <Text
      style={
        {
          "color": "#000",
          "flexShrink": 0,
          "fontSize": 10,
          "fontWeight": "bold",
        }
      }
    >
      $20
    </Text>
  </View>
  <Text
    style={
      {
        "color": "#666",
        "fontSize": 10,
        "marginTop": 4,
        "paddingHorizontal": 5,
      }
    }
  >
    Artist 1
  </Text>
</View>
`;
